encfs/compatwin.cpp

```diff
@@ -640,22 +640,44 @@ unix::stat(const char *path, struct stat_st *buffer)
     return -1;
   }
 
-
   // We need an active file handle in order to get the file index ID 
   HANDLE hff = CreateFileW(fn.c_str(), GENERIC_READ,
     FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE,
     NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
-  if (hff == INVALID_HANDLE_VALUE) {
-    errno = ERRNO_FROM_WIN32(GetLastError());
-    return -1;
-  }
-  BY_HANDLE_FILE_INFORMATION wfd;
-  if (!GetFileInformationByHandle(hff, &wfd)) {
-    errno = ERRNO_FROM_WIN32(GetLastError());
-    return -1;
-  }
-  CloseHandle(hff);
 
+  BY_HANDLE_FILE_INFORMATION hfi;
+  WIN32_FIND_DATAW wfd;
+
+  // Not sure about the default values after init, so in doubt...
+  hfi.dwFileAttributes = 0;
+  wfd.dwFileAttributes = 0;
+  hfi.nFileIndexHigh = 0;
+  hfi.nFileIndexLow = 0;
+  hfi.nFileSizeHigh = 0;
+  wfd.nFileSizeHigh = 0;
+  hfi.nFileSizeLow = 0;
+  wfd.nFileSizeLow = 0;
+  FILETIME *ftLastAccessTime = &hfi.ftLastAccessTime;
+  FILETIME *ftLastWriteTime = &hfi.ftLastWriteTime;
+  FILETIME *ftCreationTime = &hfi.ftCreationTime;
+
+  if (hff != INVALID_HANDLE_VALUE && GetFileInformationByHandle(hff, &hfi)) {
+    CloseHandle(hff);
+  }
+  else {
+    ftLastAccessTime = &wfd.ftLastAccessTime;
+    ftLastWriteTime = &wfd.ftLastWriteTime;
+    ftCreationTime = &wfd.ftCreationTime;
+    // https://bugs.ruby-lang.org/issues/6845
+    hff = FindFirstFileW(fn.c_str(), &wfd);
+    if (hff != INVALID_HANDLE_VALUE) {
+      FindClose(hff);
```

Comment:
Should we be updating the pointers to the FILETIME attributes above (e.g., to &wfd.ftLastAccessTime) as was done on line 660-662 in this case?  As-is, if GetFileInformationByHandle fails above then does it keep the file times correctly?

Reply:
Thank you for your review @jetwhiz !
If `GetFileInformationByHandle(hff, &hfi)` fails, then the pointers to the FILETIME attributes from `hfi` will be updated to those from `wfd`, computed by `FindFirstFileW(fn.c_str(), &wfd)`.
And if `FindFirstFileW` also fails, then `stat` function will return an error.
So should be OK ?
You are thinking about a specific case where times would not be correctly computed ?

Reply:
You're right, this should be fine ... in the worst case we should just be returning an error.